### ISSUE TITLE
Optimize add-on detection pings

### ISSUE THREAD

**Issue by wresuolc** [mozilla/FirefoxColor#310]
From the add-on review:

>Could we optimize the pings between the website and the add-on to detect whether the add-on is installed? Pinging the client every second sounds a bit too much to me. Instead, the website could ping the client once and if the add-on is not installed, wait for a post-install event, sent by the add-on.

**Comment by lmorchard** [comment]
TL;DR: We're doing what we're doing because we tried everything else and ran into bug after bug. This whole area of the site is loaded with race conditions and weird behavior. 

If I recall, we *did* at one point ping the add-on once and wait. But then some times, the add-on would miss the ping because it wasn't fully initialized yet. Or the add-on would send a post-install event and the site would miss it for reasons I forget. Also, the periodic pinging is a heartbeat to detect *un*-install, for which I don't think we found a reliable event. (Maybe that's changed since we started using mozAddonManager?)

**Comment by lmorchard** [comment]
Oh wait, I thought this was about the Test Pilot add-on. This is about Firefox Color. Nevermind previous comments, this is the thing here: 

We ping periodically to detect the absence and uninstall of the add-on, so we can update the site accordingly. There is no event from a webextension to signal uninstall, and this site does not have access to mozAddonManager

**Comment by wresuolc** [comment]
@wagnerand -- do you have any insight here?  Thanks!

**Comment by wagnerand** [comment]
Sorry, I missed the reply since I wasn't subscribed to this issue.

I had assumed the periodic ping was mostly to detect install, for which `runtime.onInstalled` or `runtime.onStartup` could be used. I don't know about an uninstall/disable event. Last time I asked about that, we didn't have it, but folks were interested in use-cases (hence ping @mconca and @aswan).

Either way, pinging every second sounds quite aggressive to me.

**Comment by lmorchard** [comment]
> Either way, pinging every second sounds quite aggressive to me.

Well, absent an uninstall event, that ping period is how long it takes for the site to react to uninstall. So, the longer that period gets, the slower it looks like the site takes to realize the add-on has gone away. The period is configurable with a constant, so we can change that if we think letting the site lag longer is less of a problem than a ping every second.

**Comment by aswan** [comment]
You can detect uninstall or disable with a listener for "close" in the background page, but you can't reliably do asynchronous work from that listener (since that could be misused by malicious or simply buggy extensions to prevent themselves from being disabled/uninstalled).
If a user cannot disable the extension directly from the Colors page, then Colors would have to be in a non-foreground tab when the extension is disabled, meaning that reacting quickly is probably not such a big deal...
Maybe you could check from a "focus" handler and then throttle the background check way back?

**Comment by lmorchard** [comment]
Oh, and to complicate the ping situation: We had scenarios where the add-on took longer than 1000ms to become responsive to messages, depending on individual machine performance. So we wait for [several missed pings](https://github.com/mozilla/FirefoxColor/blob/master/src/web/index.js#L36) before deciding the add-on is no longer installed. So, the response time is the ping period times max missed pings.

**Comment by lmorchard** [comment]
> You can detect uninstall or disable with a listener for "close" in the background page, but you can't reliably do asynchronous work from that listener 

Yeah, that's part of why we went with the ping hack - messages weren't reliably making it out.

**Comment by lmorchard** [comment]
> Maybe you could check from a "focus" handler and then throttle the background check way back?

We could do that, I think... What's the performance concern here that makes once per second aggressive? As in, what would a "throttled way back" period look like? If one second is a problem, does it need to be 5 seconds? 10 seconds?
